server/src/main/java/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -6942,8 +6957,9 @@ public VirtualMachine migrateVirtualMachineWithVolume(Long vmId, Host destinatio
         Pair<Host, Host> sourceDestinationHosts = getHostsForMigrateVmWithStorage(vm, destinationHost);
         Host srcHost = sourceDestinationHosts.first();
 
-        if (!isVMUsingLocalStorage(vm) && MapUtils.isEmpty(volumeToPool)
-                && (destinationHost.getClusterId().equals(srcHost.getClusterId()) || isVmVolumesOnZoneWideStore(vm))){
+        final List<VolumeVO> volumes = _volsDao.findCreatedByInstance(vm.getId());
+        if (!isAnyVmVolumeUsingLocalStorage(volumes) && MapUtils.isEmpty(volumeToPool) && destinationHost != null
+                && (destinationHost.getClusterId().equals(srcHost.getClusterId()) || isAllVmVolumesOnZoneWideStore(volumes))) {
```

Comment:
this is a complicated condition. maybe you can extract it to a method for readability?
A good name would also clarify the use of the condition and making any comments unnecessary. Something like :
```
        if (isDestinationClearAndUnambiguousFor(volumes, volumeToPool, destinationHost)) {
```

Reply:
Done